docs/plugins/debug_plugins.md

```diff
@@ -0,0 +1,290 @@
+
+Firstly, clone the code with `git clone https://github.com/seankmartin/napari-plugin-debug.git`, or download the code [here](https://github.com/seankmartin/napari-plugin-debug). Then install the package into an environment already containing napari by:
+
+```bash
+cd napari-plugin-debug/napari-simple-reload
+pip install -e .
+```
+
+### Plugin project structure
+
+```
+napari-plugin-debug/napari-simple-reload/
+├── napari_simple_reload/
+│   ├── __init__.py
+│   ├── napari.yaml
+│   └── _widget.py # We will modify this file
+├── pyproject.toml
+├── setup.cfg
+```
+
+## Using IPython to reload code during plugin development
+
+Here, we demonstrate using the [autoreload extension](https://ipython.org/ipython-doc/3/config/extensions/autoreload.html) with IPython or a Jupyter notebook to open the napari viewer with the ability to reload the changed plugin code during development.
+
+### Launch IPython and set up the viewer
+
+First, we open the napari viewer and add our three plugin widgets to the dock. Launch IPython via the `IPython` command in terminal (IPython is installed with napari) and then enter the following into the IPython console:
+
+```IPython
+In [1]: %load_ext autoreload
+
+In [2]: %autoreload 2
+
+In [3]: from napari import Viewer
+
+In [4]: from magicgui import magicgui
+
+In [5]: viewer = Viewer()
+
+In [6]: from napari_simple_reload._widget import example
+
+In [7]: viewer.window.add_plugin_dock_widget(
+    "napari-simple-reload", "Autogenerated")
+
+In [8]: viewer.window.add_plugin_dock_widget(
+    "napari-simple-reload", "Factory")
+
+In [9]: viewer.window.add_plugin_dock_widget(
+    "napari-simple-reload", "QWidget")
+```
+
+### Changing the code with IPython running
+
+Currently, clicking on the run button for any of these widgets in the napari viewer outputs "You entered **YOUR_ENTRY**!". However, we would like to change the behaviour such that a special message is printed if nothing is entered. So let's change our `example` function in `_widget.py` from the downloaded plugin code to:
+
+```Python
+def example(input_string: str) -> str:
+    output_string = (
+        f"You entered {input_string}!"
+        if input_string
+        else "Please enter something in the text box."
+    )
+    print(output_string)
+    return output_string
+```
+
+To see the updated behavior in action, just type the name of your changed function into IPython, or reference it in some way:
+
+```IPython
+# Run before changing the code
+You entered !
+
+# Run after changing the code, but before triggering an update
+You entered !
+
+In [10]: example # anything in IPython to reload that code
+
+# Now the code is reloaded, run outputs as expected
+Please enter something in the text box.
+```
```

Comment:
This code block is generating errors when compiling - I am guessing this comes from the `!` character in the code block. The following works, but I'll leave it as a suggestion, maybe there are better ones:

```suggestion
```IPython
# Run before changing the code
'You entered !'

# Run after changing the code, but before triggering an update
'You entered !'

In [10]: example # anything in IPython to reload that code

# Now the code is reloaded, run outputs as expected
Please enter something in the text box.
```
```